File: diploma-repository/src/main/java/org/arhor/diploma/repository/BaseRepository.java
package org.arhor.diploma.repository;

import org.arhor.diploma.repository.domain.DomainObject;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.NoRepositoryBean;
import org.springframework.lang.NonNull;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@NoRepositoryBean
public interface BaseRepository<T extends DomainObject<K>, K> extends JpaRepository<T, K> {

  @NonNull
  @Override
  @Transactional(readOnly = true)
  @Query("SELECT e FROM #{#entityName} e WHERE e.deleted = false")
  List<T> findAll();

  @NonNull
  @Override
  @Transactional(readOnly = true)
  @Query("SELECT e FROM #{#entityName} e WHERE e.deleted = false")
  Page<T> findAll(@NonNull Pageable pageable);

  @NonNull
  @Override
  @Transactional(readOnly = true)
  @Query("SELECT e FROM #{#entityName} e WHERE e.deleted = false")
  List<T> findAll(@NonNull Sort sort);

  @NonNull
  @Override
  @Transactional(readOnly = true)
  @Query("SELECT e FROM #{#entityName} e WHERE e.id in :ids AND e.deleted = false")
  List<T> findAllById(@NonNull Iterable<K> ids);

  @NonNull
  @Override
  @Transactional(readOnly = true)
  @Query("SELECT e FROM #{#entityName} e WHERE e.id = :id AND e.deleted = false")
  Optional<T> findById(@NonNull K id);

  @Query("SELECT e FROM #{#entityName} e WHERE e.deleted = true")
  @Transactional(readOnly = true)
  List<T> findDeleted();

  @Override
  @Transactional(readOnly = true)
  @Query("SELECT COUNT(e) FROM #{#entityName} e WHERE e.deleted = false")
  long count();

  @Override
  @Transactional(readOnly = true)
  default boolean existsById(@NonNull K id) {
    return findById(id).isPresent();
  }

  @Override
  @Modifying
  @Transactional
  @Query("UPDATE #{#entityName} e SET e.deleted = true WHERE e.id = :id")
  void deleteById(@NonNull K id);

  @Override
  @Transactional
  default void delete(T entity) {
    deleteById(entity.getId());
  }

  @Override
  @Transactional
  default void deleteAll(Iterable<? extends T> entities) {
    entities.forEach(this::delete);
  }

  @Override
  @Modifying
  @Transactional
  @Query("UPDATE #{#entityName} e SET e.deleted = true")
  void deleteAll();
}

File: diploma-server/src/main/kotlin/org/arhor/diploma/data/persist/repository/BaseRepository.kt
package org.arhor.diploma.data.persist.repository

import org.arhor.diploma.data.persist.domain.core.DomainObject
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.NoRepositoryBean
import org.springframework.transaction.annotation.Transactional
import java.io.Serializable
import java.util.*

@NoRepositoryBean
interface BaseRepository<T : DomainObject<K>, K : Serializable> : JpaRepository<T, K> {

    @Transactional(readOnly = true)
    @Query("SELECT e FROM #{#entityName} e WHERE e.isDeleted = false")
    override fun findAll(): MutableList<T>

    @Transactional(readOnly = true)
    @Query("SELECT e FROM #{#entityName} e WHERE e.isDeleted = false")
    override fun findAll(pageable: Pageable): Page<T>

    @Transactional(readOnly = true)
    @Query("SELECT e FROM #{#entityName} e WHERE e.isDeleted = false")
    override fun findAll(sort: Sort): MutableList<T>

    @Transactional(readOnly = true)
    @Query("SELECT e FROM #{#entityName} e WHERE e.id IN :ids AND e.isDeleted = false")
    override fun findAllById(ids: Iterable<K>): MutableList<T>

    @Transactional(readOnly = true)
    @Query("SELECT e FROM #{#entityName} e WHERE e.id = :id AND e.isDeleted = false")
    override fun findById(id: K): Optional<T>

    @Query("SELECT e FROM #{#entityName} e WHERE e.isDeleted = true")
    @Transactional(readOnly = true)
    fun findDeleted(): MutableList<T>

    @Query("SELECT e.id FROM #{#entityName} e WHERE e.isDeleted = true")
    @Transactional(readOnly = true)
    fun findDeletedIds(): MutableList<K>

    @Transactional(readOnly = true)
    @Query("SELECT COUNT(e) FROM #{#entityName} e WHERE e.isDeleted = false")
    override fun count(): Long

    @JvmDefault
    @Transactional(readOnly = true)
    override fun existsById(id: K): Boolean {
        return findById(id).isPresent
    }

    @Modifying
    @Transactional
    @Query("UPDATE #{#entityName} e SET e.isDeleted = true WHERE e.id = :id")
    override fun deleteById(id: K)

    @JvmDefault
    @Transactional
    override fun delete(entity: T) {
        entity.getId()?.let { deleteById(it) }
    }

    @JvmDefault
    @Transactional
    override fun deleteAll(entities: Iterable<T>) {
        deleteAllById(entities.mapNotNull { it.getId() })
    }

    @JvmDefault
    @Transactional
    override fun deleteInBatch(entities: Iterable<T>) {
        deleteAllById(entities.mapNotNull { it.getId() })
    }

    @Modifying
    @Transactional
    @Query("UPDATE #{#entityName} e SET e.isDeleted = true")
    override fun deleteAll()

    @Modifying
    @Transactional
    @Query("UPDATE #{#entityName} e SET e.isDeleted = true WHERE e.id IN :ids")
    fun deleteAllById(ids: Iterable<K>)
}